internal/ls/folding.go

```diff
@@ -0,0 +1,540 @@
+package ls
+
+import (
+	"cmp"
+	"context"
+	"regexp"
+	"slices"
+	"strings"
+	"unicode"
+
+	"github.com/microsoft/typescript-go/internal/ast"
+	"github.com/microsoft/typescript-go/internal/astnav"
+	"github.com/microsoft/typescript-go/internal/lsp/lsproto"
+	"github.com/microsoft/typescript-go/internal/printer"
+	"github.com/microsoft/typescript-go/internal/scanner"
+)
+
+func (l *LanguageService) ProvideFoldingRange(ctx context.Context, documentURI lsproto.DocumentUri) (lsproto.FoldingRangeResponse, error) {
+	_, sourceFile := l.getProgramAndFile(documentURI)
+	res := l.addNodeOutliningSpans(sourceFile)
+	res = append(res, l.addRegionOutliningSpans(sourceFile)...)
+	slices.SortFunc(res, func(a, b *lsproto.FoldingRange) int {
+		if a.StartLine != b.StartLine {
+			return cmp.Compare(a.StartLine, b.StartLine)
+		}
+		if a.StartCharacter != nil && b.StartCharacter != nil && *a.StartCharacter != *b.StartCharacter {
+			return cmp.Compare(*a.StartCharacter, *b.StartCharacter)
+		}
+		if a.EndLine != b.EndLine {
+			return cmp.Compare(a.EndLine, b.EndLine)
+		}
+		if a.EndCharacter != nil && b.EndCharacter != nil && *a.EndCharacter != *b.EndCharacter {
+			return cmp.Compare(*a.EndCharacter, *b.EndCharacter)
+		}
+		return 0
+	})
+	return lsproto.FoldingRangesOrNull{FoldingRanges: &res}, nil
+}
+
+func (l *LanguageService) addNodeOutliningSpans(sourceFile *ast.SourceFile) []*lsproto.FoldingRange {
+	depthRemaining := 40
+	current := 0
+
+	statements := sourceFile.Statements
+	n := len(statements.Nodes)
+	foldingRange := make([]*lsproto.FoldingRange, 0, 40)
+	for current < n {
+		for current < n && !ast.IsAnyImportSyntax(statements.Nodes[current]) {
+			foldingRange = append(foldingRange, visitNode(statements.Nodes[current], depthRemaining, sourceFile, l)...)
+			current++
+		}
+		if current == n {
+			break
+		}
+		firstImport := current
+		for current < n && ast.IsAnyImportSyntax(statements.Nodes[current]) {
+			foldingRange = append(foldingRange, visitNode(statements.Nodes[current], depthRemaining, sourceFile, l)...)
+			current++
+		}
+		lastImport := current - 1
+		if lastImport != firstImport {
+			foldingRangeKind := lsproto.FoldingRangeKindImports
+			foldingRange = append(foldingRange, createFoldingRangeFromBounds(
+				astnav.GetStartOfNode(findChildOfKind(statements.Nodes[firstImport],
+					ast.KindImportKeyword, sourceFile), sourceFile, false /*includeJSDoc*/),
+				statements.Nodes[lastImport].End(),
+				foldingRangeKind,
+				sourceFile,
+				l))
+		}
+	}
+
+	// Visit the EOF Token so that comments which aren't attached to statements are included.
+	foldingRange = append(foldingRange, visitNode(sourceFile.EndOfFileToken, depthRemaining, sourceFile, l)...)
+	return foldingRange
+}
+
+func (l *LanguageService) addRegionOutliningSpans(sourceFile *ast.SourceFile) []*lsproto.FoldingRange {
+	regions := make([]*lsproto.FoldingRange, 0, 40)
```

Comment:
Out of curiosity, where does the 40 initial capacity come from?

Reply:
40 is the max depth that strada used, but I don't think that we actually need to prealloc these lists.